src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+            //Here we need to Find the key of the Type, then only we will be able to find from the collection that which item in the collection
+            //corresponds to the item in delta list(Edmchangedobjectcoll).
+            Type type = typeof(TStructuralType);
+            IEnumerable<IEdmStructuralProperty> keys = EntityType.Key();
+
+            EdmChangedObjectCollection<TStructuralType> edmChangedObjectCollection = new EdmChangedObjectCollection<TStructuralType>(EntityType);
+
+            foreach (dynamic changedObj in _items)
+            {
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+                IEdmChangedObject edmChangedObject = null;
+               
+                //Get filtered item based on keys
+                TStructuralType originalObj = GetFilteredItem(type, keys, original, changedObj); ;
+
+                try
+                {
+                    IEdmDeltaDeletedEntityObject deletedObj = changedObj as IEdmDeltaDeletedEntityObject;
+
+                    if (deletedObj != null)
+                    {
+                        if (deletedObj.Reason == DeltaDeletedEntryReason.Deleted)
+                        {
+                            operation = DataModificationOperationKind.Delete;
+                        }
+                        else
+                        {
+                            operation = DataModificationOperationKind.Unlink;
+                        }
+
+                        if (originalObj != null)
+                        {
+                            //This case handle deletions
+                            original.Remove(originalObj);
+                        }
+                    }
+                    else
+                    {
+                        if (originalObj == null)
+                        {
+                            operation = DataModificationOperationKind.Insert;
+                            //This case handle additions
+                            originalObj = Activator.CreateInstance(changedObj.ExpectedClrType);
+                            original.Add(originalObj);
+                        }
+
+                        //Patch for addition/update. This will call Delta<T> for each item in the collection
+                        changedObj.Patch(originalObj);
+                    }
+
+                    edmChangedObject = changedObj;
+                }
+                catch
+                {
+                    //For handling the failed operations.
+                    HandleFailedOperation(changedObj, operation, ref edmChangedObject, originalObj);
+                }
+
+                edmChangedObjectCollection.Add(edmChangedObject);
+            }
+
+            return edmChangedObjectCollection;
+        }
+               
+
+        private void HandleFailedOperation(dynamic changedObj, DataModificationOperationKind operation, ref IEdmChangedObject edmChangedObject, TStructuralType originalObj)
```

Comment:
>ref  [](start = 104, length = 4)

why is this ref and not out?  it looks like the incoming value is always a null value.